test/object-store/sync/flx_sync.cpp

```diff
@@ -471,6 +472,79 @@ TEST_CASE("app: error handling integration test", "[sync][flx][baas]") {
     }
 }
 
+namespace {
+struct DisconnectingWebSocketInterface : sync::WebSocketInterface {
+    std::unique_ptr<WebSocketInterface> m_impl;
+    std::atomic<bool>* disconnect;
+
+    std::string_view get_appservices_request_id() const noexcept override
+    {
+        return m_impl->get_appservices_request_id();
+    }
+
+    void async_write_binary(util::Span<const char> data, sync::SyncSocketProvider::FunctionHandler&& handler) override
+    {
+        if (*disconnect) {
+            handler(Status::OK());
+        }
+        else {
+            m_impl->async_write_binary(data, std::move(handler));
+        }
+    }
+};
+
+struct DisconnectingWebSocketObserver : sync::WebSocketObserver {
+    std::unique_ptr<WebSocketObserver> m_impl;
+    std::atomic<bool>* disconnect;
+
+    void websocket_connected_handler(const std::string& protocol) override
+    {
+        m_impl->websocket_connected_handler(protocol);
+    }
+
+    void websocket_error_handler() override
+    {
+        m_impl->websocket_error_handler();
+    }
+
+    bool websocket_binary_message_received(util::Span<const char> data) override
+    {
+        if (*disconnect)
+            return true;
+        return m_impl->websocket_binary_message_received(data);
+    }
+
+    bool websocket_closed_handler(bool was_clean, sync::websocket::WebSocketError error_code,
+                                  std::string_view message) override
+    {
+        return m_impl->websocket_closed_handler(was_clean, error_code, message);
+    }
+};
+
+// A socket provider which claims to always work, but when `disconnect = true`
+// will actually drop all incoming and outgoing messages. This enables testing
+// going offline at very specfic points.
+struct DisconnectingSocketProvider : sync::websocket::DefaultSocketProvider {
```

Comment:
more of a DiscardAllTrafficSocketProvider or something rather than a DisconnectingSocketProvider since it never actually disconnects you. Is there anything in this PR that would behave differently if there was an actual disconnect where we reset all the protocol state rather than just discarding messages?

Reply:
I don't love the name, but it's "disconnecting" in the sense of disconnecting a network cable between you and the server.

Resetting the protocol state would require tying the tests to implementation details of sync connections, while this approach lets us test it via the public API.

Reply:
For these tests I do just want to be absolutely sure that no synchronization has happened after the download of the fresh realm has completed (until it's time to allow sync to happen again) and the exact details are unimportant.

Reply:
I've reworked this type to call websocket_closed_handler() (or defer the call to DefaultSocketProvider::connect()) as it turned out that dropping packets really didn't work outside of simple cases.